File: Android_Java/wit-example-ble5/app/src/main/java/com/wit/witsdk/sensor/modular/device/utils/DeviceModelFactory.java
package com.wit.witsdk.sensor.modular.device.utils;

import com.wit.witsdk.sensor.modular.device.DeviceModel;
import com.wit.witsdk.sensor.modular.device.entity.DeviceOption;
import com.wit.witsdk.sensor.modular.processor.interfaces.IDataProcessor;
import com.wit.witsdk.sensor.modular.resolver.interfaces.IProtocolResolver;

/**
 * 设备模型工厂
 *
 * @author huangyajun
 * @date 2023/2/2 13:34
 */
public class DeviceModelFactory {

    /**
     * 创建设备模型
     *
     * @author huangyajun
     * @date 2023/2/2 13:35
     */
    public static DeviceModel createDevice(String deviceName, DeviceOption deviceOption) {
        Class<? extends IDataProcessor> dataProcessorClass = deviceOption.getDataProcessorClass();
        Class<? extends IProtocolResolver> protocolResolverClass = deviceOption.getProtocolResolverClass();

        IDataProcessor iDataProcessor = null;
        IProtocolResolver iProtocolResolver = null;
        String listenerKey = deviceOption.getListenerKey();

        try {
            iDataProcessor = dataProcessorClass.newInstance();
            iProtocolResolver = protocolResolverClass.newInstance();

        } catch (Exception e) {
            e.printStackTrace();
        }

        DeviceModel deviceModel = new DeviceModel(deviceName,
                iProtocolResolver,
                iDataProcessor,
                listenerKey);
        return deviceModel;
    }
}

File: Windows_C#/Wit.Example_WTVB01BT50/WitSdk/Device/Device/Utils/DeviceModelFactory.cs
using System;
using Wit.SDK.Modular.Sensor.Device.Entity;
using Wit.SDK.Modular.Sensor.Modular.DataProcessor.Interface;
using Wit.SDK.Modular.Sensor.Modular.ProtocolResolver.Interface;

namespace Wit.SDK.Modular.Sensor.Device.Utils
{
    /// <summary>
    /// 设备模型工厂
    /// </summary>
    public class DeviceModelFactory
    {

        /// <summary>
        /// 随机设备id
        /// </summary>
        /// <returns></returns>
        public static string RandomDeviceId()
        {
            return Guid.NewGuid().ToString();
        }

        /// <summary>
        /// 从配置创建设备模型 
        /// </summary>
        /// <param name="deviceOption">设备配置</param>
        /// <returns></returns>
        public static DeviceModel CreateDeviceModel(string deviceId, string deviceName, DeviceOption deviceOption)
        {

            if (string.IsNullOrEmpty(deviceId))
            {
                throw new Exception("无法创建设备,设备没有id");
            }

            IProtocolResolver protocolResolver = null;
            IDataProcessor dataProcessor = null;

            try
            {
                if (deviceOption.ProtocolResolverType != null)
                    protocolResolver = (IProtocolResolver)deviceOption.ProtocolResolverType.Assembly.CreateInstance(deviceOption.ProtocolResolverType.FullName);
            }
            catch (Exception ex)
            {
                throw new Exception("无法创建协议解析器", ex);
            }

            try
            {
                if (deviceOption.DataProcessorType != null)
                    dataProcessor = (IDataProcessor)deviceOption.DataProcessorType.Assembly.CreateInstance(deviceOption.DataProcessorType.FullName);
            }
            catch (Exception ex)
            {
                throw new Exception("无法创建数据处理器", ex);
            }

            DeviceModel deviceModel = new DeviceModel(
                deviceId,
                deviceName,
                protocolResolver,
                dataProcessor,
                deviceOption.ListenerKey
                );
            return deviceModel;
        }


        /// <summary>
        /// 从配置创建设备模型
        /// </summary>
        /// <param name="deviceOption">设备配置</param>
        /// <returns></returns>
        public static DeviceModel CreateDeviceModel(DeviceOption deviceOption)
        {
            return CreateDeviceModel(RandomDeviceId(), "Unknown", deviceOption);
        }

        /// <summary>
        /// 创建设备模型
        /// </summary>
        /// <param name="deviceId">设备名称</param>
        /// <param name="protocolResolver">解析解析器</param>
        /// <param name="dataProcessor">数据解析器</param>
        /// <param name="dataUpdateListener">监听key</param>
        /// <returns></returns>
        public static DeviceModel CreateDeviceModel(string deviceId, string deviceName, IProtocolResolver protocolResolver, IDataProcessor dataProcessor, string dataUpdateListener)
        {
            DeviceModel deviceModel = new DeviceModel(
                deviceId,
                deviceName,
                protocolResolver,
                dataProcessor,
                dataUpdateListener
                );
            return deviceModel;
        }

    }
}
